keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivative.java

```diff
@@ -0,0 +1,177 @@
+        return partial;
+    }
+
+    public long[] getOfShape(long[] wrtShape) {
+        return Arrays.copyOfRange(partial.getShape(), 0, partial.getShape().length - wrtShape.length);
+    }
+
+    public long[] getWrtShape(long[] ofShape) {
+        return Arrays.copyOfRange(partial.getShape(), ofShape.length, partial.getShape().length);
+    }
+
+    public PartialDerivative add(PartialDerivative addition) {
+
+        if (isPresent() && addition.isPresent()) {
+            return new PartialDerivative(partial.plus(addition.partial));
+        } else if (isPresent() && addition.isEmpty()) {
+            return new PartialDerivative(getPartial());
+        } else if (isEmpty() && addition.isPresent()) {
+            return new PartialDerivative(addition.partial);
+        } else {
+            return PartialDerivative.EMPTY;
+        }
+    }
+
+    public PartialDerivative subtract(PartialDerivative subtraction) {
+
+        if (isPresent() && subtraction.isPresent()) {
+            return new PartialDerivative(partial.minus(subtraction.partial));
+        } else if (isPresent() && subtraction.isEmpty()) {
+            return new PartialDerivative(getPartial());
+        } else if (isEmpty() && subtraction.isPresent()) {
+            return new PartialDerivative(subtraction.partial.unaryMinus());
+        } else {
+            return PartialDerivative.EMPTY;
+        }
+    }
+
+    public PartialDerivative multiplyBy(double multiplier) {
+
+        if (isEmpty()) {
+            return this;
+        }
+
+        return new PartialDerivative(partial.times(multiplier));
+    }
+
+    public PartialDerivative multiplyAlongOfDimensions(DoubleTensor multiplier) {
+
+        if (isEmpty()) {
+            return this;
+        }
+
+        DoubleTensor multiplierFromLeft = increaseRankByAppendingOnesToShape(multiplier, partial.getRank());
+        DoubleTensor result = partial.times(multiplierFromLeft);
+
+        return new PartialDerivative(result);
+    }
+
+    public PartialDerivative multiplyAlongWrtDimensions(DoubleTensor multiplier) {
+
+        if (isEmpty()) {
+            return this;
+        }
+
+        DoubleTensor multiplierFromRight = increaseRankByPrependingOnesToShape(multiplier, partial.getRank());
+        DoubleTensor result = partial.times(multiplierFromRight);
+
+        return new PartialDerivative(result);
+    }
+
+    public PartialDerivative divideByAlongOfDimensions(DoubleTensor divisor) {
+
+        if (isEmpty()) {
+            return this;
+        }
+
+        DoubleTensor divisorFromLeft = increaseRankByAppendingOnesToShape(divisor, partial.getRank());
+        DoubleTensor result = partial.div(divisorFromLeft);
+
+        return new PartialDerivative(result);
+    }
+
+    public static PartialDerivative matrixMultiplyAlongOfDimensions(PartialDerivative partial, DoubleTensor multiplier, boolean partialIsLeft) {
+
+        if (partial.isEmpty()) {
+            return partial;
```

Comment:
why does this return `partial` and not `this`?

Reply:
Because it's static and there is no `this`. It's the same concept though.

Reply:
FWIW - I like this explanation.